server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -62,59 +67,96 @@ public void addServerAndUpdateSegment(
   {
     synchronized (this) {
       this.segment.set(segment);
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.computeIfAbsent(priority, p -> new HashSet<>());
+      Set<QueryableDruidServer> priorityServers;
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        priorityServers = realtimeServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      } else {
```

Comment:
There are more than two types of servers, so to make future-proof, I would add a check here that the server is `HISTORICAL`